### PR TITLE
Add mime-guess feature to statically store mimetype

### LINKED ISSUE #192: Store the result from `mime_guess` statically at compile time?
The examples all show using `mime_guess` at runtime while serving the file to guess the mimetype, but it would be nice if there were a `mime_guess` feature that would make it statically compile the result of `mime_guess` into the `EmbeddedFile` type.

### PR DESCRIPTION
**Pull request by BBaoVanC**
Closes #192

### CODE REVIEW

**Comment by pyrossh** [comment]
Hmmm wonder why the pipeline hasn't triggered but looks good though.

**Review by pyrossh — APPROVED** [review]

**Review by pyrossh — APPROVED** [review]

**Comment by pyrossh** [comment]
Will try closing and opening the PR to see if the pipeline gets triggered.

**Comment by AzureMarker** [comment]
The pipeline needs to be manually approved to run:
image

File: utils/src/lib.rs [hunk 1]
```diff
@@ -105,6 +112,12 @@ impl Metadata {
   pub fn last_modified(&self) -> Option<u64> {
     self.last_modified
   }
+
+  /// The mime type of the file
+  #[cfg(feature = "mime-guess")]
+  pub fn mimetype(&self) -> &Cow<str> {
```
Comment (AzureMarker) [inline utils/src/lib.rs:118]: This can return `&str` instead

File: impl/src/lib.rs [hunk 2]
```diff
@@ -174,6 +174,13 @@ fn embed_file(rel_path: &str, full_canonical_path: &str) -> TokenStream2 {
     Some(last_modified) => quote! { Some(#last_modified) },
     None => quote! { None },
   };
+  #[cfg(feature = "mime-guess")]
+  let mimetype = {
+    let mt = file.metadata.mimetype().to_string();
+    quote! { #mt }
```
Comment (AzureMarker) [inline impl/src/lib.rs:180]: You shouldn't need to run strings through `quote`

File: impl/src/lib.rs [hunk 3]
```diff
@@ -174,6 +174,13 @@ fn embed_file(rel_path: &str, full_canonical_path: &str) -> TokenStream2 {
     Some(last_modified) => quote! { Some(#last_modified) },
     None => quote! { None },
   };
+  #[cfg(feature = "mime-guess")]
+  let mimetype = {
+    let mt = file.metadata.mimetype().to_string();
```
Comment (AzureMarker) [inline impl/src/lib.rs:179]: `to_string` shouldn't be needed

File: impl/src/lib.rs [hunk 4]
```diff
@@ -192,7 +199,7 @@ fn embed_file(rel_path: &str, full_canonical_path: &str) -> TokenStream2 {
 
           Some(rust_embed::EmbeddedFile {
               data: std::borrow::Cow::from(bytes),
-              metadata: rust_embed::Metadata::__rust_embed_new([#(#hash),*], #last_modified)
+              metadata: rust_embed::Metadata::__rust_embed_new([#(#hash),*], #last_modified, #mimetype)
```
Comment (AzureMarker) [inline impl/src/lib.rs:202]: If you want the parameter to be optional, this call has to be extracted

**Review by AzureMarker — CHANGES_REQUESTED** [review]

File: impl/src/lib.rs (hunk 4 above)
Comment (BBaoVanC) [reply impl/src/lib.rs:202]: extracted as in it only gets added when the feature is enabled? I couldn't figure out how to do that

Comment (AzureMarker) [reply impl/src/lib.rs:202]: i.e.
```rust
let metadata_tokens = quote! {
    rust_embed::Metadata::__rust_embed_new([#(#hash),*], #last_modified, #mimetype)
};

// ...

Some(rust_embed::EmbeddedFile {
    data: std::borrow::Cow::from(bytes),
    metadata: #metadata
})
```

File: impl/src/lib.rs [hunk 5]
```diff
@@ -174,6 +174,13 @@ fn embed_file(rel_path: &str, full_canonical_path: &str) -> TokenStream2 {
     Some(last_modified) => quote! { Some(#last_modified) },
     None => quote! { None },
   };
+  #[cfg(feature = "mime-guess")]
+  let mimetype = {
+    let mt = file.metadata.mimetype().to_string();
+    quote! { #mt }
+  };
+  #[cfg(not(feature = "mime-guess"))]
+  let mimetype = quote! { None };
```
[the remark this answers is not in the document]
Comment (BBaoVanC) [reply impl/src/lib.rs:183]: Hmm somehow it did work, although I'll try and rewrite it anyways like in your other comment.

File: impl/src/lib.rs (hunk 4 above)
Comment (BBaoVanC) [reply impl/src/lib.rs]: Does this look better? https://github.com/pyrossh/rust-embed/pull/203/files#diff-8230eaab64c15b319536adf4e223a85b729dd190b4c1a2095e293fdaa4e7da57

**Review by AzureMarker — APPROVED** [review]

**Pull request merged**